pkg/controller/clusterapimachine/machine_controller.go

```diff
@@ -0,0 +1,301 @@
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Sync %q failed with %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncMachine will sync the machine with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachine(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machine")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machine")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	machine, err := c.machinesLister.Machines(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machine has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	mLog := clustoplog.WithClusterAPIMachine(c.logger, machine)
+	mLog.Debugf("found machine")
+
+	// Lookup the cluster for this machine:
+	clusterName, ok := machine.Labels[clusterNameLabel]
+	if !ok {
+		return fmt.Errorf("no cluster label set on machine")
+	}
+	cluster, err := c.clustersLister.Clusters(namespace).Get(clusterName)
+	if errors.IsNotFound(err) {
+		mLog.WithField("cluster", clusterName).Errorf("no cluster found for machine")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
+	if err != nil {
+		return err
+	}
+
+	// TODO: once cluster controller is setting resolved cluster version refs on the
+	// ClusterStatus, replace this with the resolved ref not a lookup
+	clusterVersion, err := c.cvLister.ClusterVersions(clusterSpec.ClusterVersionRef.Namespace).Get(clusterSpec.ClusterVersionRef.Name)
+	if err != nil {
+		mLog.Errorf("error looking up cluster version: %v", clusterSpec.ClusterVersionRef)
+		return err
+	}
+	/*
+		clusterStatus, err := controller.ClusterStatusFromClusterAPI(cluster)
+		if err != nil {
+			return err
+		}
+	*/
+
+	newMachine := machine.DeepCopy()
+	err = controller.PopulateMachineSpec(&newMachine.Spec, clusterSpec, clusterVersion, mLog)
+	if err != nil {
+		mLog.Errorf("error populating machine spec: %v", err)
+		return err
+	}
+
+	if !apiequality.Semantic.DeepEqual(machine, newMachine) {
+		// TODO: updating the machine here, should this be a patch operation on the template spec?
+		// NOTE: updating here immediately triggers another sync...
```

Comment:
I don't think that this note is necessary. That is overwhelmingly (if not universally) true in controllers.